### PR TITLE
Support custom forbid reason messages

### PR DESCRIPTION
**Pull request by craigfurman**
In order to allow users to communicate intent to collaborators,
optionally embed custom messages into each forbidden pattern.

The syntax is as follows: `identifier(# message goes here)?`

Example: `^fmt\.Errorf(# Please don't use this!)?$`

Regular expressions containing custom messages are effectively identical
to ones that don't, because the sub-expression containing it is marked
optional with a `?`. All this commit does is parse out any recognized
custom message, and place it prominently in the tool's output. The
regular expression itself is omitted from the tool's output when a
custom message is specified.

---

We discussed this approach in https://github.com/ashanbrown/forbidigo/issues/10. ~~I can't think of a reason a user might start their patterns with `#[`, but if they do, this PR will break that use case.~~

This should automatically work from within golangci-lint, although I haven't tested that yet. If accepted, perhaps we should add a custom message example to that project's docs?

Closes https://github.com/ashanbrown/forbidigo/issues/10.

### CODE REVIEW

File: forbidigo/forbidigo.go [hunk 1]
```diff
@@ -36,13 +37,17 @@ func (a UsedIssue) Position() token.Position {
 
 func (a UsedIssue) String() string { return toString(a) }
 
-func toString(i Issue) string {
-	return fmt.Sprintf("%s at %s", i.Details(), i.Position())
+func toString(i UsedIssue) string {
+	s := fmt.Sprintf("%s at %s", i.Details(), i.Position())
+	if i.customMsg != "" {
+		s += ": " + i.customMsg
```
Comment (craigfurman) [inline forbidigo/forbidigo.go:43]: The custom message goes on the end of the stringified issue. That seems fine to me, although I'm open to suggestion!

**Comment by ashanbrown** [comment]
Thanks @craigfurman for the PR and the thorough testcases.  I was thinking more about this and I wanted to propose a slight variation to the syntax which would look like this:

```
^fmt.\\Printf(# no extra noise allowed)?$
```

What's interesting about this syntax is that the regex remains a valid regex.  You can run it on your code (and verify the string) even without using forbidigo and it will match.  We can simply extract the capture groups and report the ones that start with '#' (or maybe just support first and last capture group).

Golang even offers a parser at https://pkg.go.dev/regexp/syntax#Parse that we can use to parse it.  See https://play.golang.org/p/KvL8L7GystE for an example of how it is parsed.

What would you think about this syntax change?

**Comment by craigfurman** [comment]
Interesting! At first glance I didn't like that comment syntax, but I have to agree that it'd be neat not to double-parse the regex, which if I understand correctly we can do using that syntax.Parse function. I'm happy to have a go at this, might be a couple of days before I get time though.

**Comment by craigfurman** [comment]
I started taking a look at this again today (apologies for the week-long latency between replies here!) and it's a little fiddlier than I was expected. We either have to make certain assumptions about where in the submatch tree the comment will be if it is present, or do a bunch of regex matching on the regex components itself, which I think defeats the point of your suggestion a little bit. I've spiked out the first approach here. Note that it doesn't actually work yet, but I think it gives the gist of what I'm saying: https://github.com/craigfurman/forbidigo/commit/b51de3b7992f9358666769eb93bc89a8b6a62171

While doing this, a thought occurred to be though: taking advantage of the fact that your proposed syntax also matches identifiers _without_ that comment, we might not actually have to write any code. We could just document this `(# comment)?` trick and call it a day! The tool's output already prints the expression used, so the comment will be present too. WDYT?

**Comment by ashanbrown** [comment]
@craigfurman It did also occur to me that we don't really need a comment if there's obviously a comment in a regex, so we could definitely start using and encouraging this syntax right away in the docs.  Parsing the comment just allows us to make the output a little more human-readable.

Thanks for taking a stab at it.  I had been imaging a recursive function that just extracted any sub-expressions that began with "#".  I suppose it would be nice if it also advised you to put a `?` after the subexpresssion, because the `'#'` symbol is never going to match anything (since `forbidigo` is focused on package names and identifiers).

**Comment by craigfurman** [comment]
Yeah I might have overthought this, extracting sub-expressions beginning with `#` is likely good enough, since that can't be contained in an identifier. How about we do that, and also replace, rather than append to, the "forbidden by pattern.." part when comments are present. That'd allow us to unconditionally use the regex provided by the user, whether it contains a comment or not, without duplicating the comment.

**Comment by craigfurman** [comment]
I've implemented what we discussed above. It's effectively a totally different implementation to what I originally posted, so I've just squashed it in and amended the message.

I've changed the output to omit the "by pattern..." part when a custom message is specified, for compactness.

Ready for your review!

File: forbidigo/patterns.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+func parse(ptrn string) (*pattern, error) {
+	p := &pattern{}
+	var err error
+	p.pattern, err = regexp.Compile(ptrn)
+	if err != nil {
+		return nil, fmt.Errorf("unable to compile pattern `%s`: %s", ptrn, err)
+	}
+	re, err := syntax.Parse(ptrn, syntax.Perl)
```
Comment (craigfurman) [inline forbidigo/patterns.go:22]: I couldn't spot a way in the stdlib to convert between a `regexp.Regexp` and a `syntax.Regexp`, so I think we might have to parse it twice, which is a bit of a shame.

File: forbidigo/patterns.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+	"strings"
+)
+
+type pattern struct {
+	pattern *regexp.Regexp
+	msg     string
+}
+
+func parse(ptrn string) (*pattern, error) {
+	p := &pattern{}
```
Comment (ashanbrown) [inline forbidigo/patterns.go:16]: just a style thing but my preference here would be to avoid dot notation and construct the final pattern at the end before we return it

File: forbidigo/forbidigo.go [hunk 4]
```diff
@@ -24,10 +24,15 @@ type UsedIssue struct {
 	identifier string
 	pattern    string
 	position   token.Position
+	customMsg  string
 }
 
 func (a UsedIssue) Details() string {
-	return fmt.Sprintf("use of `%s` forbidden by pattern `%s`", a.identifier, a.pattern)
+	explanation := fmt.Sprintf(": `%s`", a.customMsg)
```
Comment (ashanbrown) [inline forbidigo/forbidigo.go:31]: I think regular quotes rather than backticks might be more appropriate here because we are dealing with a string rather than an identifier:

```suggestion
	explanation := fmt.Sprintf(`because "%s"`, a.customMsg)
```

File: forbidigo/patterns.go (hunk 2 above)
Comment (ashanbrown) [reply forbidigo/patterns.go:22]: I agree it's not ideal but it only happens once per expression so I'm not too concerned.

**Review by ashanbrown — COMMENTED** [review]
Left a couple of small suggestions but it looks pretty good to me.

File: forbidigo/patterns_test.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+	"github.com/stretchr/testify/require"
+)
+
+func TestParseValidPattern(t *testing.T) {
+	ptrn, err := parse(`fmt\.Errorf`)
+	require.Nil(t, err)
+	assert.Equal(t, `fmt\.Errorf`, ptrn.pattern.String())
+}
+
+func TestParseValidPatternWithCustomMessage(t *testing.T) {
```
Comment (ashanbrown) [inline forbidigo/patterns_test.go:16]: If we want to be thorough here we could also add a test for a nested subgroup, ensuring the recursion works.

I personally also like the structure of having a single `TestParse` method at the top-level and writing a `t.Run()` for each subtest. A table-style test might also be appropriate here, although there aren't that many variants.

File: forbidigo/patterns.go (hunk 3 above)
Comment (craigfurman) [reply forbidigo/patterns.go:16]: I usually lean that way too (not mutating structs if it can be avoided), not sure why I went the other way here. Probably. a hangover from the first iteration of this PR.

**Comment by craigfurman** [comment]
Oops, I forgot to add the nested subgroup test. Done now.

File: forbidigo/patterns.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+	msg := extractComment(re)
+	return &pattern{pattern: ptrnRe, msg: msg}, nil
+}
+
+// Traverse the leaf submatches in the regex tree and extract a comment, if any
+// is present.
+func extractComment(re *syntax.Regexp) string {
+	for _, sub := range re.Sub {
+		if len(sub.Sub) > 0 {
+			return extractComment(sub)
```
Comment (ashanbrown) [inline forbidigo/patterns.go:33]: Sorry for not noticing this during my first review, but I think we should only return here if `extractComment` returns a non-empty string.  Otherwise, I think risk failing to return a comment if there are other sub-expressions in the pattern.

**Review by ashanbrown — CHANGES_REQUESTED** [review]
Looks good.  Just noted one more potential issue that I think it should be easy to construct a test for a fix.

File: forbidigo/patterns.go (hunk 6 above)
Comment (craigfurman) [reply forbidigo/patterns.go]: Oh, good catch! I added a commit containing a test case that indeed failed in the way I think you're describing, and the fix. Are there any other edge cases you had in mind, or does that capture it all?

**Comment by craigfurman** [comment]
Thanks for spotting that! Ready for another review, let me know if you'd prefer I squash that fix commit in first (or feel free to ask github to squash it). I've left it separate for review-ability until then.

File: forbidigo/forbidigo.go [hunk 7]
```diff
@@ -24,10 +24,15 @@ type UsedIssue struct {
 	identifier string
 	pattern    string
 	position   token.Position
+	customMsg  string
 }
 
 func (a UsedIssue) Details() string {
-	return fmt.Sprintf("use of `%s` forbidden by pattern `%s`", a.identifier, a.pattern)
+	explanation := fmt.Sprintf(` because "%s"`, a.customMsg)
```
Comment (ashanbrown) [inline forbidigo/forbidigo.go:31]:
```suggestion
	explanation := fmt.Sprintf(` because %q`, a.customMsg)
```

File: forbidigo/patterns.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse pattern `%s`: %s", ptrn, err)
+	}
+	msg := extractComment(re)
+	return &pattern{pattern: ptrnRe, msg: msg}, nil
+}
+
+// Traverse the leaf submatches in the regex tree and extract a comment, if any
+// is present.
+func extractComment(re *syntax.Regexp) string {
```
Comment (ashanbrown) [inline forbidigo/patterns.go:30]:
```suggestion
func extractComment(re syntax.Regexp) string {
```

**Review by ashanbrown — APPROVED** [review]
LGTM.  Left a couple of small suggestions

**Comment by craigfurman** [comment]
Thanks @ashanbrown! Looks like you merged this before accepting your own second suggestion though. Either way, up to you! I look forward to using this when it lands in golangci-lint.

**Comment by ashanbrown** [comment]
Thanks @craigfurman .  I think my second suggestion wasn't going to work, so I nixed it.  Thank you for your contribution.  I look forward to using it myself once it's merged into golangci-lint (which I think will happen automatically via dependabot).
